frontend/src/scenes/events/eventDefinitionsLogic.ts

```diff
@@ -0,0 +1,74 @@
+import { kea } from 'kea'
+import api from 'lib/api'
+import { EventDefinition } from '~/types'
+import { eventDefinitionsLogicType } from './eventDefinitionsLogicType'
+
+interface EventDefinitionStorage {
+    count: number
+    next: null | string
+    results: EventDefinition[]
+}
+
+export const eventDefinitionsLogic = kea<eventDefinitionsLogicType<EventDefinitionStorage, EventDefinition>>({
+    reducers: {
+        eventStorage: [
+            { results: [], next: null, count: 0 } as EventDefinitionStorage,
+            {
+                loadEventDefinitionsSuccess: (state, { eventStorage }) => {
+                    return {
+                        count: eventStorage.count,
+                        results: [...state.results, ...eventStorage.results],
+                        next: eventStorage.next,
+                    }
+                },
+            },
+        ],
+    },
+    loaders: ({ values }) => ({
+        eventStorage: [
+            { results: [], next: null, count: 0 } as EventDefinitionStorage,
+            {
+                loadEventDefinitions: async (initial?: boolean) => {
+                    const url = initial ? 'api/projects/@current/event_definitions/' : values.eventStorage.next
+                    if (!url) {
+                        throw new Error('Incorrect call to eventDefinitionsLogic.loadEventDefinitions')
+                    }
+                    return await api.get(url)
+                },
+            },
+        ],
+    }),
+    listeners: ({ actions }) => ({
+        loadEventDefinitionsSuccess: ({ eventStorage }) => {
+            if (eventStorage.next) {
+                actions.loadEventDefinitions()
+            }
+        },
+    }),
+    events: ({ actions }) => ({
+        afterMount: () => {
+            actions.loadEventDefinitions(true)
+        },
+    }),
+    selectors: {
+        loaded: [
+            // Whether *all* the event definitions are fully loaded
+            (s) => [s.eventStorage, s.eventStorageLoading],
+            (eventStorage: EventDefinitionStorage, eventStorageLoading: boolean): boolean =>
+                !eventStorageLoading && !eventStorage.next,
+        ],
+        eventDefinitions: [
+            (s) => [s.eventStorage],
+            (eventStorage: EventDefinitionStorage): EventDefinition[] => eventStorage.results,
+        ],
+        eventNames: [
+            // TODO: This can be improved for performance by enabling downstream components to use `eventDefinitions` directly and getting rid of this selector.
```

Comment:
Made the note thing that this can be improved in the future. I shunned away from updating all the downstream components now because: this PR will be large enough already, and almost all related components are not typescript yet so there's a lot of room for bugs to creep in.

Reply:
I think it's probably faster if downstream components use this selector :). Depends on what the downstream components do obviously, but if they do a lot of mapping and filtering themselves, it's easier and definitely faster to pre-calculate it all in a selector.